modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/rel/TableScanNode.java

```diff
@@ -206,14 +208,15 @@ private void requestNextBatch() {
         }
 
         if (waiting == 0) {
-            waiting = inBufSize;
+            // we must not request rows more than inBufSize
+            waiting = inBufSize - inBuff.size();
```

Comment:
why we can request next batch while **inBuff** is still not empty ?

Reply:
that's how it works right now